1.1.3.2 Connected Supabase to the project

tristan-88 opened the pull request:
- 1.1.3.2 Connected Supabase with Project.

# Set up a temporary Query to establish a connection
![Screenshot 2024-01-25 at 10 46 44 PM](https://github.com/digitalaidseattle/seattle-humane-app/assets/69748713/035bc823-5b9b-4375-b54b-04d3ab93277e)
![Screenshot 2024-01-25 at 10 47 29 PM](https://github.com/digitalaidseattle/seattle-humane-app/assets/69748713/5d65bb1c-7423-4d46-a1cc-66130f99499c)
![Screenshot 2024-01-25 at 10 47 45 PM](https://github.com/digitalaidseattle/seattle-humane-app/assets/69748713/4b0af1c5-2554-4ac1-8ad4-f6f0bed4302d)

## Note:
I think I figured out the further abstraction mentioned by Jeff and Sam but I do not believe it was necessary for me to set up for this specific change/set up just yet. 

Perhaps setting something like 
```typescript
// client/service/SupabaseService.ts
import { SupabaseClient, PostgrestResponse } from "@supabase/supabase-js";

interface SupabaseEntity {
  id: string;
  // Add other properties as needed
}

class SupabaseService<T extends SupabaseEntity> {
  private supabase: SupabaseClient;
  private tableName: string;

  constructor(supabase: SupabaseClient, tableName: string) {
    this.supabase = supabase;
    this.tableName = tableName;
  }

  async create(entity: T): Promise<T | null> {
    // TODO: Implement create method
    return null;
  }

  async update(entity: T): Promise<T | null> {
    // TODO: Implement update method
    return null;
  }

  async getById(id: string): Promise<T | null> {
    // TODO: Implement getById method
    return null;
  }

  async getAll(): Promise<T[]> {
    // TODO: Implement getAll method
    return [];
  }

  // Add more methods as needed
}

export type { SupabaseEntity };
export {  SupabaseService };

```

And we will be using client and service close to something like

```typescript

import supabaseClient from ".../utils/supabaseClient";
import { SupabaseEntity, SupabaseService } from "./client/service/SupabaseService";

// Example usage:
interface Ticket extends SupabaseEntity {
  // Define ticket properties
}

const ticketService = new SupabaseService<Ticket>(supabaseClient, "tickets");

// Now you can use ticketService.create(), ticketService.update(), etc.
...

```

I think?

samhhenderson commented:
Per your comment on implementing the abstraction - I think it's the right idea, but we'll want to name the class something that doesn't refer to supabase, like DatabaseService or something, so that if we decide later to drop supabase we can just change the implementation here instead of everywhere.

tristan-88 commented on package-lock.json:
Agree, I should have checked prior and should have realized it

DillonOLeary replied:
Why are we ignoring the package-lock.json? According to the npm documentation:

> This file is intended to be committed into source repositories

https://docs.npmjs.com/cli/v10/configuring-npm/package-lock-json

DillonOLeary commented:
It's a little difficult for me to follow what you are doing for this ticket. Did you do some kind of write up that I'm not seeing? Or would you mind adding it? While I can sorta piece it together, I'd love more explanation for why you are doing this and also explaining the screenshots. Is the John Doe stuff just mocked data? in that case it doesn't seem that we are connected to Supabase.

Additionally, I'd like to see some unit tests

jnakaso commented on README.md:
```diff
@@ -79,3 +78,6 @@ The website is built using **Next.js**, a React-based framework for server-side
 3. In order to work on Sanity components, request to be added to Sanity.io as an administrator, and then populate the respective fields in the .env.local file.
```
Supabase, not Sanity :-)

tristan-88 replied:
LOL! I did not touch or write those but I will be happy to change them if want me to add them on this push.

```
2. Open local instance in browser: <http://localhost:3000>. Note: Pages populated by Sanity.io will return 404 until Sanity is properly configured.

3. In order to work on Sanity components, request to be added to Sanity.io as an administrator, and then populate the respective fields in the .env.local file.

4. In order to work on Airtable forms, request access to Airtable and create a personal access token```


Let me know what to update these too and I can make the change and add on PR.

tristan-88 commented:
> It's a little difficult for me to follow what you are doing for this ticket. Did you do some kind of write up that I'm not seeing? Or would you mind adding it? While I can sorta piece it together, I'd love more explanation for why you are doing this and also explaining the screenshots. Is the John Doe stuff just mocked data? in that case, it doesn't seem that we are connected to Supabase.
> 
> Additionally, I'd like to see some unit tests, or maybe an explanation as to why automated testing isn't needed

@DillonOLeary Sure =)

- One is a ticket that we both can work on separately or together. 
- What I did? are the following...
### INITIAL PUSH 
1. Set up supabase by installing it in the project.
2. From the documents seems like we need the url and the api anon key
3. So I stored those in the .env.local
4. I then further added abstraction by creating supabaseClient.ts to call the keys directly and we imported the supabaseClient to files where we would invoke and interact with DB.
5. The two screenshots you see with console.logs are from a temporary test page component I created querying the test/initial table we have currently in the supabase you can see it in the supabase dashboard. I figured to establish a connection is querying from an existing table. maybe? I can be wrong. One view is on the page via browser and the second view is from my ide console.
### AFTER INITIAL PUSH 
6. I added package-lock.json in .ignore (even is not grayed out still)
7. I reverted the changes from package-lock.json to reflect all our original ones without my specific changes... ( I tried to delete the changes but it was deleting the entire file)
8. I added a .env.local.example file to add in /document the environmental variables we would need to spin up the project locally or otherwise.
9. Everything under "Note:" section is me asking what I think further abstraction will look like based on Jeff -> Sam's Suggestion and I think we concluded we did not need to set it up on this specific ticket just yet. 

#### My thoughts and questions 
- I can work on a unit test, but frankly takes me so long to create those that I do not even know how to write it for supabase/nextJS yet. 
- If you figured out how to properly connect supabase with the project you can push those changes like I said we both put our names down on this ticket. And I can remove my PR is not a big deal. It is also something we can pair into if you want, but as of right now, I feel we are connected or else I would not be able to query data from db to browser if we are not... As mentioned before I do not know is a new process and tech for me. 
- As for unit test creation, since I am a bit more used to front-end where I can do A B Testing, console.logs, and used dev tools should we create initial basic unit testing on every change? and does that mean our resident QA would create more in-depth testing?

samhhenderson reviewed (approved):
Looks good. Agreed with Dillon we will want unit testing, but in order to unblock other tasks let's move forward.

DillonOLeary commented:
Thank you for explaining that more.

> I can work on a unit test, but frankly takes me so long to create those that I do not even know how to write it for supabase/nextJS yet.

It's really important that we get unit tests working for this project, especially if it's hard to test because that might tell us there is a larger issue. It's also nice to be able to quickly identify broken logic in older code when we need to refactor.

> If you figured out how to properly connect supabase with the project you can push those changes like I said we both put our names down on this ticket. And I can remove my PR is not a big deal. It is also something we can pair into if you want, but as of right now, I feel we are connected or else I would not be able to query data from db to browser if we are not... As mentioned before I do not know is a new process and tech for me.

Would love to do some pair programming! And no need to remove your change, it seems like you were able to connect it. It would be nice to ensure that connection with integration testing though.

> As for unit test creation, since I am a bit more used to front-end where I can do A B Testing, console.logs, and used dev tools should we create initial basic unit testing on every change? and does that mean our resident QA would create more in-depth testing?

I've mostly done backend so I'm guessing we've had different experiences, but unit testing is really nice for having confirmation that your code pieces works as you intended, especially once other devs start making code changes. For this change, you are just creating a client, so I'm not sure if there is logic to test through a unit test, but we will certainly want to test this with integration testing. When you are talking about A B testing, console.logs, and dev tools are you talking about automated integration testing that happens with each PR?

And I think it's best to talk to QA about their role, I'm not sure. In my experience, devs write unit tests and integration tests.

tristan-88 commented:
@DillonOLeary  No Problem
> 

> It's really important that we get unit tests working for this project, especially if it's hard to test because that might tell us there is a larger issue. It's also nice to be able to quickly identify broken logic in older code when we need to refactor.
> 
I Agree, I probably was not expressing myself properly I was just focused on this task alone, not the whole project. I was not suggesting or even thinking it is good practice to create the whole project without any unit testing whatsoever.


> Would love to do some pair programming! And no need to remove your change, it seems like you were able to connect it. It would be nice to ensure that connection with integration testing though.
> 
I believe integration testing is contingent on unit testing. You can always pull the latest code from the sprint-1 branch since this task was merged and collaborate with the QA engineers(Justin and Yulia) on developing these tests. If time and availability allow me to join in,  I'm eager to participate as well and gain more exposure to test writing, especially as I haven't had much experience with integration tests before. 

> I've mostly done backend so I'm guessing we've had different experiences, but unit testing is really nice for having confirmation that your code pieces work as you intended, especially once other devs start making code changes. For this change, you are just creating a client, so I'm not sure if there is logic to test through a unit test, but we will certainly want to test this with integration testing. When you are talking about A B testing, console. logs, and dev tools are you talking about automated integration testing that happens with each PR?
> 
> And I think it's best to talk to QA about their role, I'm not sure. In my experience, devs write unit tests and integration tests.


Based on my experience, my primary focus has been on front-end development. In each implementation, our approach involved thorough visual and interactive testing as a preliminary step. We meticulously observed and, in some cases, conducted A/B testing to discern any version differences. For instance, we ensured that components rendered correctly and meticulously evaluated the interactive behavior, especially when data was input into a form. This process allowed us to confirm whether the observed behaviors aligned with the anticipated outcomes. Subsequently, we documented these observations and proceeded to write unit tests utilizing the appropriate testing framework.

It's worth highlighting that, in parallel to my previous experiences and yours, I've noticed a notable difference in some teams in companies where a dedicated Quality Assurance (QA) team is absent. This stands in stark contrast to the structured teams at DAS. Currently, there's some uncertainty or lack of familiarity with the proper conventions and procedures in this context.
